### ISSUE TITLE
Macros to calculate values?

### ISSUE THREAD

**Issue by p3365** [Dotneteer/spectnetide#82]
I have been experimented with macros, more specifically having macros to calculate values. Is this possible?
I have tried to create some macros to calculate values, for example:

```
BLACK   equ 0
BLUE    equ 1
RED     equ 2
MAGENTA equ 3
GREEN   equ 4
CYAN    equ 5
YELLOW  equ 6
WHITE   equ 7

Ink: macro(ink)    
  {{ink}}
.endm

Paper: macro(paper)    
  {{paper}} << 3
.endm
```

I hoped to use this like:

`LD A, Ink(BLACK) | Paper(WHITE)`

or

`DEFB Ink(BLACK) | Paper(WHITE)`

For the above I get the error:

**Error		Z0100: Unexpected token: '<<'**

Is this possible, or is there a better approach?

**Comment by Dotneteer** [comment]
The macros in the current SpectNetIde Z80 Assembler are not designed to calculate values. The assembler can use them only to emit instructions or pragmas. From the sample you provided, I see the pattern you'd like to apply. I've been thinking about adding user functions to the assembler for a while, nonetheless, I need to design this feature first.

Today you can do something like this:

```
BLACK   equ 0
BLUE    equ 1
RED     equ 2
MAGENTA equ 3
GREEN   equ 4
CYAN    equ 5
YELLOW  equ 6
WHITE   equ 7

LdColor:
  macro(reg, ink, paper)
  ld {{reg}}, {{ink}} | {{paper}} << 3
  .endm

  ; ...
  LdColor(a, BLACK, WHITE)
```

I know, it is not as clear as your code snippet, but at least, it works.

Another quick help would be to add built-in functions to the assembler for calculating `paper`, `ink` or `attribute` values. This alternative can be quickly implemented. If you have ideas of such frequently used functions, I'd be happy to discuss them.

I plan to close this issue when we have some agreement/consequences on how to proceed, so I keep the discussion open.

**Comment by p3365** [comment]
Yes, for this example, built-in functions for calculating paper ink or attribute values would work. If a general purpose use of macros to calculate values is not possible then that is OK. I suppose I was using these macros like old-style C/C++ macros rather than Z80 compiler macros.

It may be useful to add something to the documentation to say that a macro must emit an instruction?

**Comment by Dotneteer** [comment]
__Your suggestion about documentation is absolutely useful.__ Let me ask for your advise: Would you put this comment right into the overview of macros in the documentation (https://github.com/Dotneteer/spectnetide/blob/master/Documentation/Z80Assembly/Z80AssemblerReference.md#macros) or somewhere else within the Macros section?

**Comment by p3365** [comment]
Yes that sounds like a good place to put the information, thanks!

**Comment by Dotneteer** [comment]
Documentation will be fixed soon, in v1.13
